lib/SILOptimizer/Mandatory/SemanticARCOpts.cpp

```diff
@@ -26,123 +28,144 @@ using namespace swift;
 
 STATISTIC(NumEliminatedInsts, "number of removed instructions");
 
-static bool optimizeGuaranteedArgument(SILArgument *Arg,
-                                       OwnershipChecker &Checker) {
-  bool MadeChange = false;
+namespace {
 
-  // Gather all copy_value users of Arg.
-  llvm::SmallVector<CopyValueInst *, 4> Worklist;
-  for (auto *Op : Arg->getUses()) {
-    if (auto *CVI = dyn_cast<CopyValueInst>(Op->getUser())) {
-      Worklist.push_back(CVI);
-    }
-  }
+struct SemanticARCOptVisitor
+    : SILInstructionVisitor<SemanticARCOptVisitor, bool> {
+  bool visitSILInstruction(SILInstruction *i) { return false; }
+  bool visitCopyValueInst(CopyValueInst *cvi);
+  bool visitBeginBorrowInst(BeginBorrowInst *bbi);
+};
 
-  // Then until we run out of copies...
-  while (!Worklist.empty()) {
-    auto *CVI = Worklist.pop_back_val();
-
-    // Quickly see if copy has only one use and that use is a destroy_value. In
-    // such a case, we can always eliminate both the copy and the destroy.
-    if (auto *Op = CVI->getSingleUse()) {
-      if (auto *DVI = dyn_cast<DestroyValueInst>(Op->getUser())) {
-        DVI->eraseFromParent();
-        CVI->eraseFromParent();
-        NumEliminatedInsts += 2;
+} // end anonymous namespace
+
+bool SemanticARCOptVisitor::visitBeginBorrowInst(BeginBorrowInst *bbi) {
```

Comment:
[question] I thought we were using all-caps for acronyms: `BBI`? I don't actually care, but if not, there will be quite a lot of code to rewrite.